[conan] Add header-only style conan support

Fohlen opened the pull request:
Add's `conanfile.py` via [conan template](https://github.com/bincrafters/conan-templates/tree/master/header_only)

Fohlen committed:
[conan] Add header-only style conan support

Fohlen commented:
I just so copied the `.github` folder and believe it could also be deleted.

cbbowen commented on appveyor.yml:
```diff
@@ -0,0 +1,19 @@
+    PYTHON_ARCH: "32"
+
+    matrix:
+        - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
```
Even the latest MSVC isn't sufficiently standards compliant to build the Adjacency_list test.

`error C2651: 'defer<meta::v1::detail::_and_<false>::invoke,defer<meta::v1::detail::_and_<false>::invoke,std::is_same<std::integer_sequence<bool,ranges::v3::concepts::models<ranges::v3::concepts::Readable,T>::type::value,ranges::v3::concepts::models<ranges::v3::concepts::Readable,_if_<meta::v1::list<std::integral_constant<bool,B>,ranges::v3::detail::projected_<I1,P1>>,std::true_type>::type>::type::value>,std::integer_sequence<bool,ranges::v3::concepts::models<ranges::v3::concepts::Readable,T>::type::value||,ranges::v3::concepts::models<ranges::v3::concepts::Readable,_if_<meta::v1::list<std::integral_constant<bool,B>,ranges::v3::detail::projected_<I1,P1>>,std::true_type>::type>::type::value||>>,meta::v1::defer<meta::v1::invoke,F,_if_<meta::v1::list<std::integral_constant<bool,B>,ranges::v3::detail::projected_<I0,P0>>,std::true_type>::type,_if_<meta::v1::list<std::integral_constant<bool,B>,ranges::v3::detail::projected_<I1,P1>>,std::true_type>::type>>::type,ranges::v3::concepts::models<ranges::v3::concepts::CopyConstructible,C>>::type': left of '::' must be a class, struct or union`

Whereas clang works fine.

Fohlen replied:
I have removed AppVeyor for now and it is probably best if it is just re-enabled when `MSVC` can build it 👍

cbbowen commented on .github/settings.yml:
```diff
@@ -0,0 +1,42 @@
+repository:
```
Is this file necessary for CI?

cbbowen commented on test_package/conanfile.py:
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env python
```
If any of these files were copied from bincrafters/conan-templates (as you indicated this one was), the copyright needs to be retained because they're MIT licensed.

cbbowen commented on .github/settings.yml:
```diff
@@ -0,0 +1,42 @@
+  has_projects: false
+
+branches:
+  - name: "stable/*"
```
I'm not sure this library is really ready to mark a branch as `stable`.

cbbowen commented on .ci/install.sh:
```diff
@@ -0,0 +1,24 @@
+
+set -ex
+
+if [[ "$(uname -s)" == 'Darwin' ]]; then
```
I don't use a Mac.  Is it normal to have to special case everything for it?

Fohlen replied:
Yes. This is because of the following reason:

- OSX can *only* be built natively (no Docker)
- Using one `install` script is usually more comfortable

So we run the same script on multiple setups (Linux/Docker and OSX) with the effect of it is setting up a build toolchain (compiler+cmake+conan in this case).

cbbowen commented on .ci/run.sh:
```diff
@@ -0,0 +1,12 @@
+
+set -ex
+
+if [[ "$(uname -s)" == 'Darwin' ]]; then
```
See comment above regarding Mac support.

cbbowen commented on conanfile.py:
```diff
@@ -1,24 +1,41 @@
+    topics = ("graph-algorithms", "generic", "graph")
+    url = "https://github.com/cbbowen/graph/"
+    homepage = "https://github.com/cbbowen/graph/"
+    author = "Christian Bowen"
```
Just "Chris Bowen" :)

Fohlen committed:
Cleanup CI PR
- remove uneccesary GitHub settings
- remove AppVeyor for now since MSVC cannot build this
- add copyright notice where needed
- adjust author

Fohlen committed:
Forgot to add copyright to Travis config

cbbowen commented on .travis.yml:
```diff
@@ -0,0 +1,24 @@
+        env: CONAN_CLANG_VERSIONS=5.0 CONAN_DOCKER_IMAGE=conanio/clang50
+
+install:
+  - chmod +x .ci/install.sh
```
I'm still learning conan and travis, but it seems like it would be simpler (and less fraught with license complications) to do this without the shell scripts.  Something like:
```yaml
language: cpp
os: linux
dist: trusty
sudo: required
addons:
  apt:
    sources:
      - ubuntu-toolchain-r-test
    packages:
      - g++-8
env: CC=gcc-8 CXX=g++-8
before_install: pip install conan
install: conan user
script: conan create . user/channel
```

But I'm probably missing something.

Fohlen replied:
In theory this could be done. However the `xenial` and `trusty` images comes with `libstdc5` (or earlier)  and it is cumbersome and error prone to upgrade to any more recent version (which is needed for C++17). Using a pre-backed docker with all the tools makes life much easier.

cbbowen commented:
I think having Travis set up this way doesn't quite work, because it tries to fetch the latest release archive rather than using the current code, which seems to defeat the purpose of CI.  See the travis build log here: https://travis-ci.org/cbbowen/trytravis-scratch/builds/481678825.

Fohlen commented:
Yup this seems wrong, I will investigate why it does not `conan create .` locally, which would be the intended purpose.
